### PR TITLE
added ptrace system call to system monitor

### PR DESCRIPTION
**Pull request by rksharma95**
Signed-off-by: rk <ramakant@accuknox.com>

added `ptrace` system call.

Example 1:  trace cat process running directly on the host using `strace`
> Terminal A:
```
$ cat # pid 86836
```
> Terminal B
```
$ sudo su
$ strace -p 86836 # attach to cat
$ ^c # detach 
```
Telemetry Events:
```json
{"Timestamp":1657091591,"UpdatedTime":"2022-07-06T07:13:11.360036Z","ClusterName":"default","HostName":"rk-hp-pavilion-notebook","HostPID":107366,"PPID":84410,"PID":107366,"Type":"HostLog","Source":"/usr/bin/strace -p 86386","Operation":"Process","Data":"syscall=SYS_PTRACE request=PTRACE_SEIZE pid=107367","Result":"Passed","ParentProcessName":"/usr/bin/bash","ProcessName":"/usr/bin/strace","HostPPID":84410}
{"Timestamp":1657091591,"UpdatedTime":"2022-07-06T07:13:11.360990Z","ClusterName":"default","HostName":"rk-hp-pavilion-notebook","HostPID":107366,"PPID":84410,"PID":107366,"Type":"HostLog","Source":"/usr/bin/strace -p 86386","Operation":"Process","Data":"syscall=SYS_PTRACE request=PTRACE_SETOPTIONS pid=107368","Result":"Passed","ParentProcessName":"/usr/bin/bash","ProcessName":"/usr/bin/strace","HostPPID":84410}
...
[Truncated]
...
{"Timestamp":1657091593,"UpdatedTime":"2022-07-06T07:13:13.538266Z","ClusterName":"default","HostName":"rk-hp-pavilion-notebook","HostPID":107366,"PPID":84410,"PID":107366,"Type":"HostLog","Source":"/usr/bin/strace -p 86386","Operation":"Process","Resource":"/usr/bin/cat","Data":"syscall=SYS_PTRACE request=PTRACE_INTERRUPT pid=86386","Result":"Passed","ParentProcessName":"/usr/bin/bash","ProcessName":"/usr/bin/strace","HostPPID":84410}
{"Timestamp":1657091593,"UpdatedTime":"2022-07-06T07:13:13.538180Z","ClusterName":"default","HostName":"rk-hp-pavilion-notebook","HostPID":107366,"PPID":84410,"PID":107366,"Type":"HostLog","Source":"/usr/bin/strace -p 86386","Operation":"Process","Resource":"/usr/bin/cat","Data":"syscall=SYS_PTRACE request=PTRACE_DETACH pid=86386","Result":"No such process","ParentProcessName":"/usr/bin/bash","ProcessName":"/usr/bin/strace","HostPPID":84410}
```
Example 2: (unsuccessful) trace cat process running in a pod in k8s env using `strace`

> Terminal A:
```
$ cat # pid 132
```
> Terminal B
```
$ sudo su
$ strace -p 132 # attract to cat 
```
Telemetry Event/Alert:
```json
{"Timestamp":1657089880,"UpdatedTime":"2022-07-06T06:44:40.129236Z","ClusterName":"default","HostName":"rk-hp-pavilion-notebook","NamespaceName":"multiubuntu","PodName":"ubuntu-1-deployment-6fcd9cbc4f-wkqw7","ContainerID":"51c0c58fb05b0602b030485b9545e800a628f46a0db44b73c4f90fea47cc8457","ContainerName":"ubuntu-1-container","HostPID":80537,"PPID":102,"PID":182,"PolicyName":"DefaultPosture","Type":"MatchedPolicy","Source":"/usr/bin/strace -p 132","Operation":"Process","Data":"syscall=SYS_PTRACE request=PTRACE_SEIZE pid=183","Action":"Block","Result":"Permission denied","ContainerImage":"kubearmor/ubuntu-w-utils:0.1@sha256:b4693b003ed1fbf7f5ef2c8b9b3f96fd853c30e1b39549cf98bd772fbd99e260","ParentProcessName":"/bin/bash","ProcessName":"/usr/bin/strace","HostPPID":10769}
{"Timestamp":1657089880,"UpdatedTime":"2022-07-06T06:44:40.129711Z","ClusterName":"default","HostName":"rk-hp-pavilion-notebook","NamespaceName":"multiubuntu","PodName":"ubuntu-1-deployment-6fcd9cbc4f-wkqw7","ContainerID":"51c0c58fb05b0602b030485b9545e800a628f46a0db44b73c4f90fea47cc8457","ContainerName":"ubuntu-1-container","HostPID":80537,"PPID":102,"PID":182,"PolicyName":"DefaultPosture","Type":"MatchedPolicy","Source":"/usr/bin/strace -p 132","Operation":"Process","Data":"syscall=SYS_PTRACE request=PTRACE_ATTACH pid=132","Action":"Block","Result":"Operation not permitted","ContainerImage":"kubearmor/ubuntu-w-utils:0.1@sha256:b4693b003ed1fbf7f5ef2c8b9b3f96fd853c30e1b39549cf98bd772fbd99e260","ParentProcessName":"/bin/bash","ProcessName":"/usr/bin/strace","HostPPID":10769}
```

### CODE REVIEW

**Comment by rksharma95** [comment]
PTAL @nyrahul @nam-jaehyun

**Comment by nyrahul** [comment]
Please rebase. Can we target this for v0.6 ? CC @achrefbensaad

File: KubeArmor/monitor/syscallParser.go [hunk 1]
```diff
@@ -192,6 +193,81 @@ func readSockaddrFromBuff(buff io.Reader) (map[string]string, error) {
 	return res, nil
 }
 
+// getPtraceReq
+func getPtraceReq(req uint32) string {
```
Comment (achrefbensaad) [inline KubeArmor/monitor/syscallParser.go:231]: Can we change this to use switch instead ?

File: KubeArmor/monitor/logUpdate.go [hunk 2]
```diff
@@ -191,6 +191,36 @@ func (mon *SystemMonitor) UpdateLogs() {
...
+				}
+
+				var request string
+				var pid string
+
+				if val, ok := msg.ContextArgs[0].(string); ok {
+
+					request = val
+				}
+				var pid_ uint32
```
Comment (achrefbensaad) [inline KubeArmor/monitor/logUpdate.go]: Can you please address the lint error

File: KubeArmor/types/types.go [hunk 3]
```diff
@@ -514,6 +514,9 @@ type SELinuxRule struct {
 // PidMap for host pid -> process node
 type PidMap map[uint32]PidNode
 
+// PidMap for pid -> host pid
```
Comment (achrefbensaad) [inline KubeArmor/types/types.go]: Can you please address the lint error ?

**Review by achrefbensaad — DISMISSED** [review]

**Comment by achrefbensaad** [comment]
Please squash your commits to one.

File: KubeArmor/BPF/system_monitor.c [hunk 4]
```diff
@@ -1530,3 +1534,14 @@ if (sk_lingertime_offset - gso_max_segs_offset == 2)
 }
 
 char LICENSE[] SEC("license") = "Dual BSD/GPL";
+
+// ptrace syscall
+int syscall__ptrace(struct pt_regs *ctx){
+    if (skip_syscall())
+        return 0;
```
Comment (achrefbensaad) [inline KubeArmor/BPF/system_monitor.c:1702]: you need to add the required annotations as we migrated from bcc.

File: KubeArmor/BPF/system_monitor.c [hunk 5]
```diff
@@ -1530,3 +1547,5 @@ if (sk_lingertime_offset - gso_max_segs_offset == 2)
 }
 
 char LICENSE[] SEC("license") = "Dual BSD/GPL";
+
+
```
Comment (achrefbensaad) [inline KubeArmor/BPF/system_monitor.c]: Please remove the extra lines.

File: KubeArmor/BPF/system_monitor.c [hunk 6]
```diff
@@ -1529,4 +1546,4 @@ if (sk_lingertime_offset - gso_max_segs_offset == 2)
     return 0;
 }
 
-char LICENSE[] SEC("license") = "Dual BSD/GPL";
+char LICENSE[] SEC("license") = "Dual BSD/GPL";
```
Comment (achrefbensaad) [inline KubeArmor/BPF/system_monitor.c:1717]: Please add a new line here

File: KubeArmor/BPF/system_monitor.c [hunk 7]
```diff
@@ -1634,4 +1651,4 @@ if (sk_lingertime_offset - gso_max_segs_offset == 2)
...
```
Comment (Ankurk99) [inline KubeArmor/BPF/system_monitor.c:1717]: please remove the newline at the end of the file

File: KubeArmor/monitor/systemMonitor.go [hunk 8]
```diff
@@ -287,7 +294,7 @@ func (mon *SystemMonitor) InitBPF() error {
 	mon.Logger.Print("Initialized the eBPF system monitor")
 
 	// sysPrefix := bcc.GetSyscallPrefix()
-	systemCalls := []string{"open", "openat", "execve", "execveat", "socket", "connect", "accept", "bind", "listen", "unlink", "unlinkat", "rmdir", "chown", "setuid", "setgid", "fchownat"}
+	systemCalls := []string{"open", "openat", "execve", "execveat", "socket", "connect", "accept", "bind", "listen", "unlink", "unlinkat", "rmdir", "ptrace", "rmdir", "chown", "setuid", "setgid", "fchownat"}
```
Comment (Ankurk99) [inline KubeArmor/monitor/systemMonitor.go]: there is one extra "rmdir"

**Comment by rksharma95** [comment]
@Ankurk99 

`[PASS] Successfully tested KubeArmor`
runtime tests is successful on my system. tested on ubuntu 20.04 with k3s env.

**Comment by Ankurk99** [comment]
@rksharma95 Please squash all the commits.

File: KubeArmor/BPF/system_monitor.c (hunk 7 above)
Comment (Ankurk99) [reply KubeArmor/BPF/system_monitor.c:1717]: my bad, a new line should be added at the end of file

**Review by Ankurk99 — APPROVED** [review]
LGTM!

**Review by achrefbensaad — APPROVED** [review]
LGTM

File: KubeArmor/monitor/systemMonitor.go [hunk 9]
```diff
@@ -515,6 +526,10 @@ func (mon *SystemMonitor) TraceSyscall() {
 					continue
 				}
 
+			} else if ctx.EventID == SysPtrace {
+				if len(args) != 2 {
+					continue
+				}
```
Comment (daemon1024) [inline KubeArmor/monitor/systemMonitor.go]: this seems to be duplicated, already added in 508-511

File: KubeArmor/monitor/logUpdate.go [hunk 10]
```diff
@@ -301,6 +301,36 @@ func (mon *SystemMonitor) UpdateLogs() {
...
+					pid = strconv.Itoa(int(val))
+				}
+
+				if msg.ContainerID == "" {
+					log.Resource = mon.GetCommand(msg.ContainerID, traceePid)
+
+				} else {
+					ActivePidProcMap := *(mon.ActivePidProcMap)
+					hostPid := ActivePidProcMap[msg.ContainerID][traceePid]
+					log.Resource = mon.GetCommand(msg.ContainerID, hostPid)
```
Comment (daemon1024) [inline KubeArmor/monitor/logUpdate.go]: Can you document why we need to do this? what is traceePid? why do we keep a seperate map to track it?

**Review by daemon1024 — CHANGES_REQUESTED** [review]
Minor nits. Please find them inline.

Can you add an explanation for why did we need a separate map to handle it?
Also seems like the ARM PR is conflicting this, you will need to rebase and retest things.

File: KubeArmor/monitor/logUpdate.go (hunk 10 above)
Comment (rksharma95) [reply KubeArmor/monitor/logUpdate.go]: tracee pid is the pid of the process that is being traced or the tracer process is attached to. the map is to keep track of pids per container basis as the argument passed with ptrace syscall will give pid number as per container namespace.

**Comment by rksharma95** [comment]
Currently the proposed changes introduces a new map to keep track of the pid->hostPid mapping per container basis. it seems quite unfeasible and to avoid maintaining a separate map we can take any of the two approaches:

1. traversing the existing map using loop to find required host pid mapping.
2. (preferred imo) lsm hook [ptrace_access_check](https://elixir.bootlin.com/linux/v6.0.7/source/security/apparmor/lsm.c#L1206) can be used to get required mete data (full path in this case.) of the tracee process.

the later approach handles everything in kernel space and the hook is available in all the targeted kernel versions. 

@daemon1024 wdyt?

File: KubeArmor/monitor/logUpdate.go (hunk 10 above)
Comment (rksharma95) [reply KubeArmor/monitor/logUpdate.go]: we've moved from this logic, now using lsm-hook instead of maintaining a separate map.

**Comment by daemon1024** [comment]
LGTM. Can you share a sample telemtry event? Thanks

**Comment by rksharma95** [comment]
Example (Host)log:
```
== Log / 2022-11-14 08:07:30.716251 ==
ClusterName: default
HostName: hp
Type: HostLog
Source: /usr/bin/strace test
Resource: /usr/bin/test
Operation: Process
Data: syscall=SYS_PTRACE request=PTRACE_SYSCALL pid=47211 process=/usr/bin/test
Result: Passed
HostPID: 47208
HostPPID: 47207
PID: 47208
PPID: 47207
ParentProcessName: /usr/bin/sudo
ProcessName: /usr/bin/strace

```
Example Alert:
```
== Alert / 2022-11-14 08:10:25.847611 ==
ClusterName: default
HostName: hp
NamespaceName: syscalls
PodName: ubuntu-1-deployment-66448fc84f-k7c7j
Labels: container=ubuntu-1
ContainerName: ubuntu-1-container
ContainerID: 41c0fafea867f4dd536060b3424aa3af960cad96020152d8e12d7ec0449168db
ContainerImage: docker.io/kubearmor/ubuntu-w-utils:0.1@sha256:b4693b003ed1fbf7f5ef2c8b9b3f96fd853c30e1b39549cf98bd772fbd99e260
Type: MatchedPolicy
PolicyName: DefaultPosture
Source: /usr/bin/strace ls
Resource: /usr/bin/strace
Operation: Process
Action: Block
Data: syscall=SYS_PTRACE request=PTRACE_SEIZE pid=187 process=/usr/bin/strace
Enforcer: AppArmor
Result: Permission denied
HostPID: 47433
HostPPID: 47344
PID: 186
PPID: 172
ParentProcessName: /bin/bash
ProcessName: /usr/bin/strace
```

**Comment by daemon1024** [comment]
We support blocking Syscalls?

**Comment by rksharma95** [comment]
> We support blocking Syscalls?

ig yes https://docs.kubearmor.com/kubearmor/getting-started/security_policy_specification#syscalls

**Comment by daemon1024** [comment]
> For System calls monitoring, we only support audit mode no matter what the action is set to.

But we only did audit?

**Comment by rksharma95** [comment]
> But we only did audit?

yes it seems currently only audit is supported.

**Comment by daemon1024** [comment]
Then why do you have a Permission denied alert? 😅

**Comment by rksharma95** [comment]
> Then why do you have a Permission denied alert? sweat_smile

ig it's because sys_ptrace capability is not assigned by default by the container runtime.

and we get alerts for denials outside KubeArmor right!

output of /var/log/kern.log shows
```
Nov 14 15:09:35 hp kernel: [20774.373235] audit: type=1400 audit(1668418775.003:133642): apparmor="DENIED" operation="ptrace" profile="kubearmor-syscalls-ubuntu-1-deployment-ubuntu-1-container" pid=58432 comm="strace" requested_mask="trace" denied_mask="trace" peer="kubearmor-syscalls-ubuntu-1-deployment-ubuntu-1-container"
```

**Comment by daemon1024** [comment]
Okay. Makes sense. But the ptrace observer we adding. Shouldn't it show up with `Operation: Syscall`? provided ptrace is allowed.

**Comment by rksharma95** [comment]
> Okay. Makes sense. But the ptrace observer we adding. Shouldn't it show up with `Operation: Syscall`? provided ptrace is allowed.

Operation is not Syscall because there's no policy applied to Audit the Syscall.

**Comment by daemon1024** [comment]
Okay, then can you apply a sample policy and share the relevant telemetry event for the same?

**Comment by rksharma95** [comment]
currently we are not auditing ptrace syscall [Audit Syscalls](https://github.com/kubearmor/KubeArmor/blob/main/KubeArmor/monitor/syscallParser.go#L773), 
but i tested it with local changes adding ptrace to syscall list and i'm getting the alert with `Operation: Syscall`. 
```
== Alert / 2022-11-16 08:32:51.354132 ==
ClusterName: default
HostName: hp
Type: MatchedHostPolicy
PolicyName: hsp-hp-audit-syscall-ptrace
Severity: 5
Source: /usr/bin/strace ls
Resource: /usr/bin/strace
Operation: Syscall
Action: Audit
Data: syscall=SYS_PTRACE request=PTRACE_SEIZE pid=61177 process=/usr/bin/strace
Result: Permission denied
HostPID: 61176
HostPPID: 61175
PID: 61176
PPID: 61175
ParentProcessName: /usr/bin/sudo
ProcessName: /usr/bin/strace

```

**Comment by daemon1024** [comment]
> currently we are not auditing ptrace syscall [Audit Syscalls](https://github.com/kubearmor/KubeArmor/blob/main/KubeArmor/monitor/syscallParser.go#L773)

Is this added here? Changes to audit ptrace?

**Comment by rksharma95** [comment]
> > currently we are not auditing ptrace syscall [Audit Syscalls](https://github.com/kubearmor/KubeArmor/blob/main/KubeArmor/monitor/syscallParser.go#L773)
> 
> Is this added here? Changes to audit ptrace?

no should it be added?

**Comment by daemon1024** [comment]
I believe so.

**Comment by rksharma95** [comment]
added ptrace to audit syscalls, tested locally
```== Alert / 2022-11-25 08:12:26.709259 ==
ClusterName: default
HostName: hp
NamespaceName: multiubuntu
PodName: ubuntu-3-deployment-7cc5cf69f7-ztnzc
Labels: container=ubuntu-3,group=group-1
ContainerName: ubuntu-3-container
ContainerID: d2efaf88f36049a20b36a8685159c51019dac5faf09fe26e49a154b3e207f573
ContainerImage: docker.io/kubearmor/ubuntu-w-utils:0.1@sha256:b4693b003ed1fbf7f5ef2c8b9b3f96fd853c30e1b39549cf98bd772fbd99e260
Type: MatchedPolicy
PolicyName: audit-all-ptrace
Severity: 3
Source: /usr/bin/strace ls
Resource: /bin/ls
Operation: Syscall
Action: Audit
Data: syscall=SYS_PTRACE request=PTRACE_SYSCALL pid=201 process=/bin/ls
Result: Passed
HostPID: 32994
HostPPID: 30495
PID: 199
PPID: 155
ParentProcessName: /bin/bash
ProcessName: /usr/bin/strace
```

**Review by daemon1024 — APPROVED** [review]

**Pull request merged**